### PR TITLE
[deps!] Upgraded Django and dropped support for Python 3.6 #117

### LINKED ISSUE #96: [cleanup] Find out what to do with compatibility shims for ipaddress library
The coverage build shows that the compatibility shims for ipaddress library are not being executed anymore:
https://coveralls.io/jobs/78517731/source_files/7752510585

Maybe we can remove those, but before doing so we have to understand why this is happening. Maybe the ipaddress library has been updated on recent minor versions updates of python 3.6?

### LINKED ISSUE #102: [bug] Get next available IP doesn't work well with /32 on IPv4
Try creating a /32 subnet on ipv4, then try creating an address, select the newly create subnet and you'll see an alert saying "No IP address available", but this alert is wrong, because saving the address for the /32 works.

### LINKED ISSUE #117: [deps] Add support for django 3.2 and django 4.0 alpha
Let's make sure this module can run with Django 3.2 and Django 4.0 alpha.

### LINKED ISSUE #120: [ci] Add python 3.8 and python 3.9, remove python 3.6
Add python 3.8 and python 3.9, remove python 3.6 from github actions testing.

Fix any bug that may appear.

### PR DESCRIPTION
**Pull request by pandafy**
- Dropped support for Python 3.6
- Added support for Django 3.2 and Django 4.0
- Replaced xlrd with openpyxl
- Bumped django-reversion~=4.0.1

Closes #96
Closes #102
Closes #120
Closes #117

**Backward incompatible change**: Subnets with /32 prefixlen
can provision an IP address same as the network address.

### CODE REVIEW

**Review by nemesisdesign — CHANGES_REQUESTED** [review]
There's some conflicts with master.

File: openwisp_ipam/base/models.py [hunk 1]
```diff
@@ -212,13 +225,10 @@ def _read_ipaddress_data(self, reader, subnet):
...
-            book = xlrd.open_workbook(file_contents=file.read())
-            sheet = book.sheet_by_index(0)
-            row = []
-            for row_num in range(sheet.nrows):
-                row.append(sheet.row_values(row_num))
-            reader = iter(row)
+        if file.name.endswith(('.xlsx')):
+            book = openpyxl.load_workbook(filename=file)
+            sheet = book.worksheets[0]
+            reader = sheet.values
```
Comment (pandafy) [inline openwisp_ipam/base/models.py:231]: Earlier the project was using [xlrd](https://xlrd.readthedocs.io/en/latest/changes.html#id1) package for reading xls and xlsx files. But the latest version of xlrd now only processes xls files. Hence, I swapped the package for [openxlpy](https://openpyxl.readthedocs.io/) which can read xlsx files (but not xls files). I changed the dependency with the thought that more and more people will be using xlsx files in the current time. 

We can look for other alternatives if we really want to support xls files.

File: openwisp_ipam/base/models.py [hunk 2]
```diff
@@ -143,7 +144,16 @@ def _validate_master_subnet_consistency(self):
...
-        for host in self.subnet.hosts():
+        # NOTE: Shim for Python 3.7
+        # In Python < 3.8, subnet.hosts() does not include network
+        # address if prefixlen of subnet is 32.
+        # See https://bugs.python.org/issue28577.
+        py_major_ver, py_minor_ver, _, _, _ = sys.version_info[:]
+        if self.subnet.prefixlen == 32 and py_major_ver == 3 and py_minor_ver < 8:
+            subnet_hosts = [self.subnet.network_address]
+        else:
+            subnet_hosts = self.subnet.hosts()
```
Comment (nemesisdesign) [inline openwisp_ipam/base/models.py:155]: I think this needs a dedicated test.

File: openwisp_ipam/base/models.py (hunk 1 above)
Comment (nemesisdesign) [reply openwisp_ipam/base/models.py:231]: if we can do csv that's fine

Comment (pandafy) [reply openwisp_ipam/base/models.py:231]: Right now we can do CSV and XLSX. 

The library for parsing CSV is ships with Python. (no external library/package is required)
For parsing XLSX we will be using [openxlpy](https://openpyxl.readthedocs.io/)

With this PR, we will be dropping support for XLS. Should we also drop support for XLSX?

Comment (nemesisdesign) [reply openwisp_ipam/base/models.py:229]: it's fine as it is, let's not touch this further right now, there's no reason to.

**Review by nemesisdesign — APPROVED** [review]

**Pull request merged**